### ISSUE TITLE
Specify SSH Key for Private Ansible Galaxy Role

### ISSUE THREAD

**Issue by AndrewPaglusch** [ansible-semaphore/semaphore#897]
I am running the `latest` tag (as of March 18, 2022) of Semaphore in Docker. 

My repository contains a `requirements.yml` file so that Ansible Galaxy can pull in roles and collections from Galaxy and GitHub. One of the roles listed in that file can only be accessed via a SSH key, as it's a private GitHub repository.

**requirements.yml**
```
---
roles:
  - name: common
    src: git+ssh://git@github.com/AndrewPaglusch/Common-Ansible-Role.git
[...]
```

Here's the portion of the task logs that shows the Galaxy install failing because of permission issues:
```
8:08:35 PM    Starting galaxy role install process
8:08:35 PM    [WARNING]: - common was NOT installed successfully: - command /usr/bin/git
8:08:35 PM    clone ssh://git@github.com/AndrewPaglusch/Common-Ansible-Role.git common failed
8:08:35 PM    in directory /tmp/semaphore/.ansible/tmp/ansible-local-133s9seyhlk/tmppl7lvn4n
8:08:35 PM    (rc=128) - Cloning into 'common'... Host key verification failed. fatal: Could
8:08:35 PM    not read from remote repository. Please make sure you have the correct access
8:08:35 PM    rights and the repository exists.
8:08:35 PM    ERROR! - you can use --ignore-errors to skip failed roles and finish processing the list.
8:08:35 PM    Running galaxy failed: exit status 1
```

I have not been able to find a way to specify the SSH key that ansible-galaxy will use when installing roles and requirements. I have tried adding the `Common-Ansible-Role` repo to Semaphore (along with a deployment key), but that didn't seem to help at all.

I also generated a ed25519 keypair and added it to my GitHub account (no longer using a deployment key). I added the public key to Semaphore and set that to be used for accessing the main playbook repository. It doesn't look like this key is used for accessing the Galaxy roles though.

I see the `ssh_config_path` option in the configuration, but I was not able to find any documentation about this option. Perhaps this can be used to fix the problem I'm running into?

In summary, my question: Is there a way to tell Semaphore what SSH key to use for pulling role/collections from private Git repositories?

**Comment by AndrewPaglusch** [comment]
Is there any more information I can provide to assist someone in helping me find a solution to this problem?

**Comment by RyanGannon** [comment]
> Is there any more information I can provide to assist someone in helping me find a solution to this problem?

I don't suppose you found a solution to this?

**Comment by AndrewPaglusch** [comment]
Unfortunately, no. I still have not been able to find a solution to this.

**Comment by nlvw** [comment]
@AndrewPaglusch As long as you run semaphore as a non-root user you can setup `~/.ssh` with custom passwordless ssh keys.  Setup the `~/.ssh/config` to use your key when talking to the git server specified in your requirements.yml file and register the keys under a read-only service account on your git server.

That's the workaround we used.

**Comment by AndrewPaglusch** [comment]
@nlvw Great idea! I hadn't considered that approach. I will give that a shot and report back. Thank you very much.

**Comment by AndrewPaglusch** [comment]
@nlvw Thank you again for your help. That solved my problem. Here's a small write-up with more details on how I did this for others who are struggling with the same problem:

I generated a public/private SSH keypair for Semaphore to use when accessing my private GitHub repo:
```
$ ssh-keygen -t ed25519 -f semaphore
```

I mapped in the SSH private key and a SSH config file into the correct locations inside the Semaphore container:
```
[...]
  semaphore:
    container_name: semaphore
    image: ansiblesemaphore/semaphore:latest
    volumes:
      - ./semaphore/data:/etc/semaphore
      - ./semaphore/ssh-config/ssh-config:/home/semaphore/.ssh/config:ro # <<<<<
      - ./semaphore/ssh-config/keys/semaphore:/home/semaphore/.ssh/semaphore.key:ro # <<<<<
      - /root/ansible:/ansible
[...]
```

I changed the ownership and mode of the files like so:
```
$ chown -R 1001:1001 /opt/docker/semaphore/ssh-config
$ chmod 0600 /opt/docker/semaphore/ssh-config/ssh-config # <- ssh config
$ chmod 0600 /opt/docker/semaphore/ssh-config/keys/semaphore # <- private key
```

I added the public key as a deployment key to the private repo on GitHub.

Here are the contents of the ssh config (/opt/docker/semaphore/ssh-config/ssh-config). This tells Semaphore to use the mapped-in private key when talking to GitHub
```
Host *
    StrictHostKeyChecking no

Host github.com
  HostName github.com
  IdentityFile /home/semaphore/.ssh/semaphore.key
  User git
```

Now Semaphore can clone down private repos from GitHub!

**Comment by nlvw** [comment]
This shouldn't be closed as the above is a work around and not native support for the full capability of Ansible Galaxy (private git repo support).

**Comment by AndrewPaglusch** [comment]
Agreed. I'll open this back up.

**Comment by JacobCoffee** [comment]
@AndrewPaglusch did you happen to figure this out?

**Comment by AndrewPaglusch** [comment]
@JacobCoffee Unfortunately not. The workaround mentioned earlier is the only way I've been able to get this to work.

**Comment by ipimpat** [comment]
This not working is bit of bummer, hope it can be solved

**Comment by rwaffen** [comment]
i also stumbled upon this. wanted to separate my local roles/-folder into git modules, but no, you cannot do this without the workaround. sad_panda_face.jpg. thought i specified a ssh key to clone the ansible-code repo from which i run my playbooks, why not used it down there with the galaxy? 😞

**Comment by hmoffatt** [comment]
Maybe this works with the latest beta which has the key added to an ssh agent?

**Comment by PhilKet** [comment]
> Maybe this works with the latest beta which has the key added to an ssh agent?

Because I have the same problem, I just gave the 2.9.39-beta a try. But it seems like the issue still persists.
It would be great if this could be looked into in context with the switch to ssh-agent, as this is still a major drawback.

**Comment by rossm-mf** [comment]
same here. definitely need to find a solution here

**Comment by rome-legacy** [comment]
for me this apporach works. running semaphore in a k8s container.
i have generated an ssh key pair
mounted private key as `/home/semaphore/.ssh/id_rsa` and public key as `/home/semaphore/.ssh/id_rsa` into the container and the authentication was successful. you also need to check the permissions as required by ssh for private keys (https://superuser.com/questions/215504/permissions-on-private-key-in-ssh-folder). i didn't even create any ssh configs, just used default key names.

**Comment by ShadowJonathan** [comment]
currently also running into this, unfortunately
